packages/swing-store-lmdb/check-lmdb.js

```diff
@@ -1,14 +1,14 @@
 import fs from 'fs';
-import {
-  openSwingStore as openSwingStoreLMDB,
-  initSwingStore as initSwingStoreLMDB,
-} from '@agoric/swing-store-lmdb';
+import { assert, details as X } from '@agoric/assert';
+
 import {
   openSwingStore as openSwingStoreSimple,
   initSwingStore as initSwingStoreSimple,
 } from '@agoric/swing-store-simple';
-
-import { assert, details as X } from '@agoric/assert';
+import {
```

Comment:
Hmm.. I like having a non-`cosmic-swingset` place for `getBestSwingStore`, but it feels weird for one kind of swingstore to depend upon another. This is the best non-independent place for it (we don't want swingset to depends upon swing-store-lmdb, and certainly swing-store-simple shouldn't either).

If we aren't going to add a new package for this (which feels overkill), let's at least make it clear that swing-store-lmdb is taking on the responsibility for exporting this, so we don't accidentally remove it under maintenance. I think there's a `package.json` way to enumerate the exported modules, right?

I'm not thrilled with adding "deep exports" to packages in the first place, although they they don't seem too bad for auxilliary things like helper functions. But I definitely want to stay away from a policy where absolutely any source file within the package is a potential import that must be maintained at its exact location forever. (Especially given how Endo wants to enumerate the exports of each package, rather than go searching to resolve every incoming import).

Reply:
@FUDCo you might want to be aware of this too, swing-store-lmdb growing an extra export name.

Reply:
This is scary.  What is this for?  It doesn't seem like a good idea to transparently comingle persistent and ephemeral stores.

I also strongly share @warner's concern about dependency; if you are going to have something that incorporates both kinds of stores I think needs to be in a separate package from either of them.

Reply:
> This is scary. What is this for?

It is code that's been around for quite some time.  It is shared between the on-chain swingset and the ag-solo, this PR is only to move it to the lmdb package.  We needed it for WSL compatibility, since on certain Windows platforms LMDB would fail the checks in this file, so it required falling back to use the simple swing store.

@dtribble and @warner (who I believe authored this code) know more about this.  Our platform works reasonably well on Windows already, so I'm loathe to throw it out unless there's a better solution.

> It doesn't seem like a good idea to transparently comingle persistent and ephemeral stores.

It doesn't.  swing-store-simple is persistent when used as it is in this file.  Read it in [its entirety](https://github.com/Agoric/agoric-sdk/tree/mfig-2945-separate-ag-solo/packages/swing-store-lmdb/check-lmdb.js) you'll see that the simple store is used in a persistent manner.

> In particular, the simple store is intended as a testing tool while the lmdb store is intended as a production tool.

For a production tool, lmdb is not available on all platforms.  Even though it was intended as a testing tool, simple store is useful for a non-performant but portable implementation.

> I also strongly share @warner's concern about dependency; if you are going to have something that incorporates both kinds of stores I think it needs to be in a separate package from either of them.

The LMDB store is not usable in cross-platform production without this fallback.  I think that merits its inclusion in this package.  Also, when the fallback is no longer necessary, this is a good place to insist that LMDB is both supported and used exclusively, without having to adjust all of its callers.

I mostly wanted to get this code out of cosmic-swingset since `packages/solo` is gradually evolving to something more flexible.  If the current factoring is so undesirable that you would block the landing of this PR, so be it.  A little piece of me dies every time I have to create a minuscule fresh package when its functionality is needed by many (all?) production users of an existing package.

Reply:
That was my rant.  I'm now sufficiently upset by the conflict that I'll just move it back to `package/cosmic-swingset`.